### PR TITLE
server: Create new api for table metadata

### LINKED ISSUE #128894: Create new HTTP API for retrieving table metadata
In order to improve performance on the DB console databases tables page, a new API should be create to fetch table metadata from the new `system.table_metadata` table. This API should support the following:

- pagination via pageSize and pageNum query string parameters
- sorting via a sortBy and sortOrder query string parameters
- Filtering via dbId, storeId and name query string parameters
    - dbId will filter for all tables belonging to provided dbId.
    - multiple storeIds are supported. This return tables in ANY of the provided storeIds
    - name will filter based on table name

 
This API should return a list of table metadata results based on the query string parameters and paginationInfo so the front-end can build a paginated table as necessary

Jira issue: CRDB-41262

Epic CRDB-37558

### PR DESCRIPTION
**Pull request by kyle-a-wong**
adds a new api v2 endpoint, `/api/v2/table_metadata/`.

This new API enables the ability to retrieve paginated table metadata for a given database.

Resolves: https://github.com/cockroachdb/cockroach/issues/128894
Epic: [CRDB-37558](https://cockroachlabs.atlassian.net/browse/CRDB-37558)
Release note: None

### CODE REVIEW

File: pkg/server/apiutil/apiutil.go [hunk 1]
```diff
@@ -30,3 +32,59 @@ func WriteJSONResponse(ctx context.Context, w http.ResponseWriter, code int, pay
...
+			}
+			queryArgInts = append(queryArgInts, i)
+		}
+
+		return QueryStringVal[[]int]{Value: queryArgInts, Exists: true}
+	}
+	return QueryStringVal[[]int]{Exists: false}
+}
+
+type QueryStringVal[T any] struct {
```
Comment (angles-n-daemons) [inline pkg/server/apiutil/apiutil.go:87]: Do you feel strongly about adding this struct? I feel that nil or zero-value checks should be done to indicate existence, eg:

```go
if c.tableName == nil {
```
or
```go
if c.tableName == "" {
```

File: pkg/server/api_v2_databases_metadata.go [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+	qs := struct {
+		dbId      apiutil.QueryStringVal[int]
+		tableName apiutil.QueryStringVal[string]
+		storeIds  apiutil.QueryStringVal[[]int]
+		sortBy    apiutil.QueryStringVal[string]
+		sortOrder apiutil.QueryStringVal[string]
+		pageNum   apiutil.QueryStringVal[int]
+		pageSize  apiutil.QueryStringVal[int]
+	}{
+		dbId:      apiutil.GetIntQueryStringVal(queryValues, dbIdKey),
```
Comment (angles-n-daemons) [inline pkg/server/api_v2_databases_metadata.go:105]: What are your thoughts about inlining the utility functions into a function for this struct?

```go
type tableMetadataQueryArgs  struct {
     dbId            int
     tableName string
     ...
}

func newTableMetadataQueryArgs(vals url.Values) (tableMetadataQueryArgs, error) {
    // if you want to check zero values
    dbIdStr := vals.Get(dbIdKey)
    dbId, err := strconv.Atoi(dbIdStr)
    if err != nil {
        return errors.Wrapf(err, "could not parse tableMetadata query arg %s", dbIdKey)
    }

     // if nil checks work better to verify existence
    var tableName *string
    if vals.Has(nameKey) {
        name := vals.Get(nameKey)
        tableName = &name
    }

    storeIds := make([]int, len(vals[storeIds])
    for _, storeIdStr := range vals[storeIds] {
         storeId, err := strconv.Atoi(storeIdStr)
         if err != nil {
             return errs.Wrapf(err, "could not parse tableMetadata storeId: %s", storeIdStr)
         }
         storeIds = append(storeIds)
    }
   

    return tableMetadataQueryArgs{
        tableName: tableName,
        ...
    }
} 
```

It's a bit wordy, but I think more clearly defines how the query will be parsed (absent an api specification)

**Review by angles-n-daemons — COMMENTED** [review]
couple of comments, this looks pretty good overall

File: pkg/server/api_v2_databases_metadata.go [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+
+// GetTableMetadata returns a paginated response of table metadata and statistics. This is not a live view of
+// the table data but instead is cached data that had been precomputed at an earlier time.
+//
+// The user making the request will receive table metadata based on the CONNECT database grant and admin privilege.
+// If the user provides a database id that they are not authorized for, they will receive an empty response. Similarly,
+// if the user does not provide a database id, the result set will only include databases in which they have the
+// previously mentioned authorizations.
+//
+// This endpoint supports the following query parameters:
```
[the remark this answers is not in the document]
Comment (angles-n-daemons) [reply pkg/server/api_v2_databases_metadata.go]: Oh I remember this discussion, definitely makes sense - feel free to proceed in either direction

File: pkg/server/api_v2_databases_metadata.go [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+// name: storeId
+// type: integer
+// description: The id of the store to filter tables by. If the table contains data the store, it will be included in
+// the result set. Multiple storeId query parameters are support. If multiple are provided, a table will be included in
+// the result set if it contains data in at least one of the stores.
+func (a *apiV2Server) GetTableMetadata(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	ctx = a.sqlServer.AnnotateCtx(ctx)
+	sqlUser := authserver.UserFromHTTPAuthInfoContext(ctx)
+	// TODO (kyle): build http method handling directly into route registration
```
[the remark this answers is not in the document]
Comment (angles-n-daemons) [reply pkg/server/api_v2_databases_metadata.go:112]: got it, I see that makes sense to me

File: pkg/server/api_v2_databases_metadata.go (hunk 2 above)
Comment (angles-n-daemons) [reply pkg/server/api_v2_databases_metadata.go]: I see, yeah it makes sense since they're passed individually as arguments to the next function down - I think my only question is around the utility functions `GetIntQueryStringVal`. Do you have any thoughts about trying thing PR without them?

Comment (kyle-a-wong) [reply pkg/server/api_v2_databases_metadata.go]: Done

Comment (kyle-a-wong) [reply pkg/server/api_v2_databases_metadata.go]: Actually, I added GetIntQueryStringVal back in. I think not having this function results in a lot of duplicated code to get the param value and convert it to an int.

**Review by xinhaoz — COMMENTED** [review]
Nice! Question on the virtual table usage but looking great overall!

> Reviewed 4 of 120 files at r3, 2 of 11 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128231#-:-O6Nx4Ai6_MOi5FgYe14:b9qyx5s)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @angles-n-daemons, @dhartunian, and @kyle-a-wong)

___
*[`-- commits` line 4 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/128231#-O5TobPk5C2_Dw0kkB3j):*
I think this message can be polished a bit.
___
*[`pkg/server/api_v2_databases_metadata.go` line 145 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/128231#-O5TrRsnE7GCO18C3K0c:-O5TrRsnE7GCO18C3K0d:bh7k7im) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea5105adf00fb28d2d0beff99d2fbeb6297d88e0/pkg/server/api_v2_databases_metadata.go#L145)):*
> ```Go
> 		}
> 
> 		switch qs.sortOrder.Value {
> ```

nit: Why do we need a switch here?
```
sortOrder := "ASC"
if qs.sortOrder.Value == "desc" {
  sortOrder = "DESC"
}
```
___
*[`pkg/server/api_v2_databases_metadata.go` line 180 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/128231#-O6MKlEg1j-hibmAwi0x:-O6MKlEg1j-hibmAwi0y:bwv9wht) ([raw file](https://github.com/cockroachdb/cockroach/blob/efbf2dbad82304b4b4625020d9c7044596c055ee/pkg/server/api_v2_databases_metadata.go#L180)):*
> ```Go
> 	var sortOrder string
> 	if sortBy != "" {
> 		switch sortOrderQS {
> ```

nit: Do we need a switch here?
```
var sortOrder string
if sortBy != "" {
    sortOrder = "ASC"
    if sortOrderQS == "desc" {
        sortOrder = "DESC"
    }
}
```
___
*[`pkg/server/api_v2_databases_metadata.go` line 252 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/128231#-O6MO3rg2Q2UYVwGnmJt:-O6MO3rg2Q2UYVwGnmJu:b3t2it8) ([raw file](https://github.com/cockroachdb/cockroach/blob/efbf2dbad82304b4b4625020d9c7044596c055ee/pkg/server/api_v2_databases_metadata.go#L252)):*
> ```Go
> 			count(*) OVER() as total_row_count
> 		FROM system.table_metadata tbm
> 		JOIN crdb_internal.databases dbs ON dbs.id = tbm.db_id
> ```

Do we know if the usage of this virtual table scales well to large clusters / won't hit perf issues with a lot of descriptors?

**Review by kyle-a-wong — COMMENTED** [review]
> Reviewed 2 of 7 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128231)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @angles-n-daemons, @dhartunian, and @xinhaoz)

___
*[`pkg/server/api_v2_databases_metadata.go` line 180 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/128231#-O6MKlEg1j-hibmAwi0x:-O6WRtxy0BGbBJ7IGfsE:b-n9tkhd) ([raw file](https://github.com/cockroachdb/cockroach/blob/efbf2dbad82304b4b4625020d9c7044596c055ee/pkg/server/api_v2_databases_metadata.go#L180)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

nit: Do we need a switch here?
```
var sortOrder string
if sortBy != "" {
    sortOrder = "ASC"
    if sortOrderQS == "desc" {
        sortOrder = "DESC"
    }
}
```
</blockquote></details>

changed
___
*[`pkg/server/api_v2_databases_metadata.go` line 252 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/128231#-O6MO3rg2Q2UYVwGnmJt:-O6WRwbh-Jjlu0KXkFPU:b-hw79p1) ([raw file](https://github.com/cockroachdb/cockroach/blob/efbf2dbad82304b4b4625020d9c7044596c055ee/pkg/server/api_v2_databases_metadata.go#L252)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

Do we know if the usage of this virtual table scales well to large clusters / won't hit perf issues with a lot of descriptors?
</blockquote></details>

I removed the usage of this virtual table as it isn't strictly necessary.  One thing that this extra join provided was protection against "losing" tables in the view in the case that they have been renamed, since the name will no longer match the cluster_database_privileges. But I'm not sure how frequent an action like this occurs, but I assume it isn't very frequent any probably isn't worth the extra join.

To answer the question, I'm not sure about the performance of this VT in large clusters.

**Review by xinhaoz — APPROVED** [review]
:lgtm: Thanks for addressing comments! Let's get these APIs merged to start connecting to the UI. I'm sure we'll need to iterate on them slightly anyways once that's in progress.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128231#-:-O6WbVAV3nBwP3B8LZyL:b-mykh1r)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @angles-n-daemons, @dhartunian, and @kyle-a-wong)

___
*[`-- commits` line 2 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/128231#-O6WbPhq3SJxYXu9usfZ):*
supernit: We also typically don't capitalize the title after the `pkgName:` portion.

**Review by angles-n-daemons — APPROVED** [review]
:lgtm: also lgtm!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128231#-:-O6Wkfxr4Xz2WwPZ-SV1:b-s8fz3x)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @dhartunian, @herkolategan, @kyle-a-wong, @renatolabs, and @xinhaoz)

**Comment by kyle-a-wong** [comment]
TFTR!

bors r+
